Add basic file drag and drop support

ShadowMitia opened the pull request:
Adds basic drag and drop support to bevy.

It simply exposes the relevant events already found in winit. 
This has been tested on Linux, and because it simply wraps around the winit events, it should work on other systems just fine (hopefully!).

I still want to try wasm/web before merging. (Or someone else, I haven't played with wasm in rust or bevy much yet!)

Let me know if there's anything missing!

smokku commented on Cargo.toml:
```diff
@@ -165,6 +165,10 @@ path = "examples/app/return_after_run.rs"
 path = "examples/app/thread_pool_resources.rs"
 
+[[example]]
+name = "draganddrop"
```
Can we keep examples sorted alphabetically?

ShadowMitia replied:
I never realised they were in alphabetical order...

smokku commented on crates/bevy_winit/src/lib.rs:
```diff
@@ -350,6 +351,35 @@ pub fn winit_runner(mut app: App) {
                     }
                 }
+                WindowEvent::DroppedFile(path_buf) => {
+                    let mut cursor_left_events =
```
It should be called `dropped_file_events`, right?

ShadowMitia replied:
Indeed. Also fixed this in the other functions.

smokku commented on crates/bevy_winit/src/lib.rs:
```diff
@@ -370,6 +400,7 @@ pub fn winit_runner(mut app: App) {
                 );
                 app.update();
             }
+
```
Other matches are not separated.

ShadowMitia replied:
I'm not sure what you're saying here sorry, could you explain please?

DJMcNab commented:
Could you clarify in the title that this is file drag and drop support?

ShadowMitia commented:
@DJMcNab Will do. I'll update the commit message as well.

ShadowMitia commented:
I've tested on wasm, and this doesn't work unfortunently. But it looks like winit doesn't have it either https://github.com/rust-windowing/winit/blob/master/FEATURES.md#input-handling-1.

cart commented on Cargo.toml:
```diff
@@ -133,6 +133,11 @@ path = "examples/3d/z_sort_debug.rs"
 path = "examples/app/custom_loop.rs"
 
+[[example]]
+name = "draganddrop"
```
Can we rename the example / file to "drag_and_drop"? I like to use snake_case for all file names.

cart commented on crates/bevy_winit/src/lib.rs:
```diff
@@ -350,6 +351,35 @@ pub fn winit_runner(mut app: App) {
+                }
+                WindowEvent::HoveredFile(path_buf) => {
+                    let mut hovered_file_events =
+                        app.resources.get_mut::<Events<HoveredFile>>().unwrap();
```
I think we should probably combine all drag and drop events into a single event. It seems like order of operations matters here (ex: when did the cancel happen).

(not relevant to this pr, but there are a couple of other events here that we should revisit when it comes to combining/ordering, such as CursorLeft/CursorEntered)

ShadowMitia commented:
I took a stab at creating a simple plugin and resource to handle dropped files (and added it to `DefaultPlugins).
Is this the idea you had in mind @cart? I'm not compeltely satisfied with the interface, but then I'm still kinda new at Rust.

The plugin gives you a `DragAndDrop` resource, which you can iterate using a `drain` method.
Each file is a tuple of `(PathBuf, and WindowId)`. If a path is a directory, the user is left to manage what to do.

I can't see the point of allowing direct access to `HoveredFiles` for now so I didn't create an interface for it, but maybe I'm wrong.
There is an `is_hovered` method if you want to do something special during hovering. (I'm thinking changing the cursor icon mainly or have some kind of feedback).

ShadowMitia commented:
For reference, I'm also attempting to add file drag and drop into winit https://github.com/rust-windowing/winit/pull/1808

cart commented:
I was more talking about combining the drag and drop events into a single event stream to preserve ordering (see the changes I just pushed).

I think the drag and drop plugin introduces a bit too much complexity for what it adds, so I reverted that in favor of the consolidated event.

cart commented:
Just rebased on master / resolved conflicts. I think this is good to go!

ShadowMitia commented:
I fixed the path in the README and changed the system name in the example to something a bit clearer.
Otheriwse I think it's all good (untill I get wasm working anyway, if it happens :grin: )
